File: 백준/Silver/11403. 경로 찾기/경로 찾기.cc
#include <bits/stdc++.h>
using namespace std;

int n;
int arr[105][105];
bool visited[105];

bool dfs(int start, int end) {
  fill(visited, visited + n + 2, false);
  stack<int> s;
  s.push(start);

  while(!s.empty()) {
    int cur = s.top();
    s.pop();

    if(visited[cur]) {
      continue;
    }

    if(cur != start) {
      visited[cur] = 1;
    }

    for(int next=0; next<n; next++) {
      if(visited[next]) {
        continue;
      }
      if(arr[cur][next]) {
        if(next == end) {
          return true;
        }
        s.push(next);
      }
    }
  }
  return false;
}

int main(void)
{
  ios::sync_with_stdio(0);
  cin.tie(0);
  cout.tie(0);

  cin >> n;

  for(int i=0; i<n; i++) {
    for(int j=0; j<n; j++) {
      cin >> arr[i][j];
    }
  }

  for(int i=0; i<n; i++) {
    for(int j=0; j<n; j++) {
      cout << dfs(i, j) << " ";
    }
    cout << "\n";
  }

  return 0;
}

File: 백준/Silver/11403. 경로 찾기/경로 찾기.java
import java.util.Scanner;
import java.util.Stack;

public class Main {
    static int n;
    static int[][] board;
    static boolean[] visited;

    static boolean dfs(int st, int en) {
        for(int i=0; i<n+1; i++) {
            visited[i] = false;
        }

        Stack<Integer> s = new Stack<>();
        s.add(st);

        while(!s.empty()) {
            int cur = s.peek();
            s.pop();

            if(visited[cur]) {
                continue;
            }

            if(cur != st) {
                visited[cur] = true;
            }

            for(int next=0; next<n; next++) {
                if(visited[next]) {
                    continue;
                }

                if(board[cur][next] == 1) {
                    if(next == en) {
                        return true;
                    }
                    s.add(next);
                }
            }
        }
        return false;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        board = new int[n+1][n+1];
        visited = new boolean[n+1];

        for(int i=0; i<n; i++) {
            visited[i] = false;
            for(int j=0; j<n; j++) {
                board[i][j] = sc.nextInt();
            }
        }

        for(int i=0; i<n; i++) {
            for(int j=0; j<n; j++) {
                if(dfs(i, j)) {
                    System.out.print("1 ");
                } else {
                    System.out.print("0 ");
                }
            }
            System.out.println();
        }
    }
}
